keps/sig-release/0028-k8s-image-promoter.md

```diff
@@ -0,0 +1,102 @@
+---
+kep-number: 28
+title: Image Promoter
+authors:
+  - "@javier-b-perez"
+owning-sig: sig-release
+participating-sigs:
+  - TBD
+reviewers:
+  - TBD
+  - "@listx"
+approvers:
+  - "@thockin"
+editor: TBD
+creation-date: 2018-09-05
+last-updated: 2018-09-05
+status: draft
+---
+
+# Image Promoter
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+* [Proposal](#proposal)
+  * [Staging Container Registry](#staging-container-registry)
+  * [Production Container Registry](#production-container-registry)
+  * [Promotion Process](#promotion-process)
+* [Graduation Criteria](#graduation-criteria)
+* [Infrastructure Needed](#infrastructure-needed)
+
+
+## Summary
+
+For security reasons, we cannot allow everyone to publish container images into the official kubernetes container registry. This is why we need a process that allows us to review who built an image and who approve it to be shown in the official channels.
+
+
+## Motivation
+
+There are multiple reasons why we should have a process to publish container images in place:
+
+* We cannot allow all community members to publish images into the official kubernetes container registry.
+* We should restrict who can push images to a small set of members and some systems accounts for automation.
```

Comment:
We cannot control individual images or directories in GCR. This is why I say it should be a "small set of members".
We can restrict who can merge or approve PR into the manifest's repository, then leave the rest to the promoter tool). So those who approve PRs don't even need to have write access into the container registry.

Reply:
@javier-b-perez Would it make sense to have each sub-project get a new staging area of their own, and have the promoter pull from any of those?

I guess we'd need a new GCP project for each bucket (I can't see how to make >1 bucket in a project).  The promoter works on hash, so it should Just Work, right?

Reply:
@thockin it is possible to have a per project staging container registry, that means we need to create a GCP project per each GCR we want to handle. 
The promoter tool just need read access to this GCR and it can copy the container image with the same digest from project staging to official container registry.

How do you see the staging domain working? something like:
staging-k8s.gcr.io/project1/imageA -> GCR/project1/imageA
staging-k8s.gcr.io/project2/imageB -> GCR/project2/imageB
(not sure if this is possible, might not)

Reply:
I'm just thinking out loud.  The options are really 

a) Be careful who can push because anyone who can write can do anything.

b) Give each sub-team their own space.

e.g. gcr.io/k8s-foo-staging for project foo and gcr.io/k8s-bar-staging for project bar, both promote to `gcr.k8s.io/{foo,bar}/{foo,bar}:<tag>`

@dims @mkumatag thoughts?